### PR TITLE
feat: Add date component

### PR DESCRIPTION
**Pull request by paulgain**
# Date component
A new date component that allows default and custom validation of dates.

### CODE REVIEW

File: src/forms/elements/__stories__/FieldDate.stories.jsx [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+addDecorator(withKnobs)
+
+const ERRORS = {
+  DAY: 'Enter a valid Epoch day',
+  MONTH: 'Enter a valid Epoch month',
+  YEAR: 'Enter a valid Epoch year',
+}
+
+storiesOf('Forms', module)
+  .add('FieldDate - default validation', () => (
```
Comment (rafenden) [inline src/forms/elements/__stories__/FieldDate.stories.jsx]: I suggest merging this two stories into one.

File: src/forms/elements/FieldDate.jsx [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+import {
+  BORDER_WIDTH_FORM_ELEMENT_ERROR,
+  SPACING,
+} from '@govuk-react/constants'
+
+import FieldWrapper from './FieldWrapper'
+import useField from '../hooks/useField'
+import { useFormContext } from '../..'
+
+const DATE_FORMAT = 'YYYY-MM-DD'
```
Comment (rafenden) [inline src/forms/elements/FieldDate.jsx]: I suggest we call it `VALIDATION_DATE_FORMAT` so it's clear that this is used only for validation.

File: src/forms/elements/FieldDate.jsx [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  return (
+    <FieldWrapper {...{ name, label, hint, error }}>
+      <StyledInputWrapper error={error}>
+        {error && <ErrorText>{error}</ErrorText>}
+        <StyledList>
+          <StyledLabel>
+            <LabelText>{labels.day}</LabelText>
+            <Input
+              id={`${name}-day`}
+              name={`${name}-day`}
```
Comment (rafenden) [inline src/forms/elements/FieldDate.jsx]: I gave some though on that recently and perhaps we should use dots here, e.g. `address.postcode` so in the future we could convert (on submission) form values from flat to complex object:
```
{
  values: {
    address.postcode: "SE1 123"
  }
}
```
becomes
```
{
  values: {
    address: {
       postcode: "SE1 123"
    }
  }
}
```
which will spare us some transformations

File: src/forms/elements/FieldDate.jsx [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+          </StyledLabel>
+        </StyledList>
+      </StyledInputWrapper>
+    </FieldWrapper>
+  )
+}
+
+FieldDate.propTypes = {
+  name: PropTypes.string.isRequired,
+  label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
```
Comment (rafenden) [inline src/forms/elements/FieldDate.jsx]: We can just use `PropTypes.node`

File: src/forms/elements/__stories__/FieldDate.stories.jsx (hunk 1 above)
Comment (paulgain) [reply src/forms/elements/__stories__/FieldDate.stories.jsx]: I purposely split them so it's clear to developers you are given default date validation out of the box, meaning, you don't have to roll your own unless the default implementation doesn't fit your requirements.

File: src/forms/elements/FieldDate.jsx (hunk 2 above)
Comment (paulgain) [reply src/forms/elements/FieldDate.jsx]: 👍

File: src/forms/elements/FieldDate.jsx (hunk 3 above)
Comment (paulgain) [reply src/forms/elements/FieldDate.jsx]: 👍

File: src/forms/elements/FieldDate.jsx (hunk 4 above)
Comment (paulgain) [reply src/forms/elements/FieldDate.jsx]: 👍

File: src/forms/elements/__stories__/FieldDate.stories.jsx (hunk 1 above)
Comment (rafenden) [reply src/forms/elements/__stories__/FieldDate.stories.jsx]: Cool, it makes sense.

**Pull request merged**